src/group_theory/group_action/End_hom.lean

```diff
@@ -0,0 +1,29 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import category_theory.types
+import category_theory.endomorphism
```

Comment:
Maybe we should move `End_hom` somewhere much lower in the imports? It would match how we have `monoid.End` etc much earlier

Reply:
Sorry, how do you mean?

Reply:
Currently the monoid structure on `X -> X` is deduced from the category structure on `Type`. I think this makes sense. But it would also make sense to define it earlier, like `equiv.perm X` (which is just the group of units in `category_theory.End X`).

Reply:
I think it would definitely make sense to define the monoid structure earlier; as you say, we do it super early for `equiv.perm`, `monoid.End`, `mul_aut`, `module.End`, etc. It's odd that the most basic structure in this family is the one that pulls in the most importa.

Reply:
To be clear, what I'm suggesting is that you just add these lines to `logic.function.basic`:
```lean
/-- The monoid of endomorphisms.

See also `category_theory.End` for a generalization of this idea. -/
def function.End (α : Type*) := α → α

instance (α : Type*) : monoid (function.End α) :=
{ one := id,
  mul := (∘),
  mul_assoc := λ f g h, rfl,
  mul_one := λ f, rfl,
  one_mul := λ f, rfl, }

instance (α : Type*) : inhabited (function.End α) := ⟨1⟩
```

Reply:
Sounds good. `logic.function.basic` doesn't know what `monoid` is, so I tentatively put all this in `group_theory/group_action/defs` (this file is all about the functor represented by `function.End \a`!), but wouldn't mind moving to another file.